transports/quic/src/connection.rs

```diff
@@ -0,0 +1,423 @@
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+mod connecting;
+mod substream;
+
+use crate::{
+    endpoint::{self, ToEndpoint},
+    Error,
+};
+pub use connecting::Connecting;
+pub use substream::Substream;
+use substream::SubstreamState;
+
+use futures::{channel::mpsc, ready, FutureExt, StreamExt};
+use futures_timer::Delay;
+use libp2p_core::muxing::{StreamMuxer, StreamMuxerEvent};
+use parking_lot::Mutex;
+use std::{
+    any::Any,
+    collections::HashMap,
+    net::SocketAddr,
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll, Waker},
+    time::Instant,
+};
+
+/// State for a single opened QUIC connection.
+#[derive(Debug)]
+pub struct Connection {
+    /// State shared with the substreams.
+    state: Arc<Mutex<State>>,
+    /// Channel to the [`endpoint::Driver`] that drives the [`quinn_proto::Endpoint`] that
+    /// this connection belongs to.
+    endpoint_channel: endpoint::Channel,
+    /// Pending message to be sent to the [`quinn_proto::Endpoint`] in the [`endpoint::Driver`].
+    pending_to_endpoint: Option<ToEndpoint>,
+    /// Events that the [`quinn_proto::Endpoint`] will send in destination to our local
+    /// [`quinn_proto::Connection`].
+    from_endpoint: mpsc::Receiver<quinn_proto::ConnectionEvent>,
+    /// Identifier for this connection according to the [`quinn_proto::Endpoint`].
+    /// Used when sending messages to the endpoint.
+    connection_id: quinn_proto::ConnectionHandle,
+    /// `Future` that triggers at the [`Instant`] that [`quinn_proto::Connection::poll_timeout`]
+    /// indicates.
+    next_timeout: Option<(Delay, Instant)>,
+}
+
+impl Connection {
+    /// Build a [`Connection`] from raw components.
+    ///
+    /// This function assumes that there exists a [`Driver`](super::endpoint::Driver)
+    /// that will process the messages sent to `EndpointChannel::to_endpoint` and send us messages
+    /// on `from_endpoint`.
+    ///
+    /// `connection_id` is used to identify the local connection in the messages sent to
+    /// `to_endpoint`.
+    ///
+    /// This function assumes that the [`quinn_proto::Connection`] is completely fresh and none of
+    /// its methods has ever been called. Failure to comply might lead to logic errors and panics.
+    pub(crate) fn from_quinn_connection(
+        endpoint_channel: endpoint::Channel,
+        connection: quinn_proto::Connection,
+        connection_id: quinn_proto::ConnectionHandle,
+        from_endpoint: mpsc::Receiver<quinn_proto::ConnectionEvent>,
+    ) -> Self {
+        debug_assert!(!connection.is_closed());
```

Comment:
I hit this point at random when doing my testing (not local test so its not easy to reproduce). Will this debug assert remain in place?

Reply:
Oh can you provide your setup? It would be nice for us to write a test and catch this case.

Reply:
It happens at random when querying (or even bootstrapping) the DHT (either in rust-ipfs or just using libp2p itself with the bootstrap nodes). I would probably need to grab the logs from libp2p and see if that show anything useful before the assert which i can do later on today.

Reply:
Sorry for the late response. I did put a small code up on https://github.com/dariusc93/libp2p-quic-test, which will produce that panic at random. Will be putting providing a log here shortly.

Reply:
@kpp I uploaded the log to https://raw.githubusercontent.com/dariusc93/libp2p-quic-test/master/libp2p-quic-test.log.2022-11-05-01 . The panic seems to happen around when the connection is closed with `EndpointDriverCrashed`

Reply:
I guess my main question would be is there a point where a connection being closed would be a problem when being passed into `from_quinn_connection`? I can monitor logs with the release profile (which while running havent seen any problems so far with connecting to peers, rather that be locally or over the network), but just curious about the assert there. From what I caught so far from the logs is it happens when the endpoint crash (which happens right after an application error) with only quic, but i guess then why is it when it had a successful connection prior. I could test without relay too or maybe slim down the transport as a means of testing

Reply:
Thank you! I think this is enough info for us to fix the bug.

Reply:
Thank you for the detailed bug report @dariusc93! Much appreciated. 
I removed the `debug_assert!(!is_closed())` in 7770943. `quinn-proto` sets the state of a new inbound connection directly to `Drained` in some cases (depending on the frames received in the initial packet), `is_closed` is then true. Thus this is not a bug in the implementation, but the assert statement was incorrect.